pkg/store/prometheus.go

```diff
@@ -545,42 +552,54 @@ func (p *PrometheusStore) LabelNames(ctx context.Context, r *storepb.LabelNamesR
 		}
 	}
 
+	if len(lbls) > 0 {
+		for _, lbl := range extLset {
+			lbls = append(lbls, lbl.Name)
+		}
+		sort.Strings(lbls)
+	}
+
 	return &storepb.LabelNamesResponse{Names: lbls}, nil
 }
 
 // LabelValues returns all known label values for a given label name.
 func (p *PrometheusStore) LabelValues(ctx context.Context, r *storepb.LabelValuesRequest) (*storepb.LabelValuesResponse, error) {
-	externalLset := p.externalLabelsFn()
+	if r.Label == "" {
+		return nil, status.Error(codes.InvalidArgument, "label name parameter cannot be empty")
+	}
+
+	extLset := p.externalLabelsFn()
 
 	// First check for matching external label which has priority.
-	if l := externalLset.Get(r.Label); l != "" {
+	if l := extLset.Get(r.Label); l != "" {
 		return &storepb.LabelValuesResponse{Values: []string{l}}, nil
 	}
 
+	match, matchers, err := matchesExternalLabels(r.Matchers, extLset)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	if !match {
+		return &storepb.LabelValuesResponse{Values: nil}, nil
+	}
+
 	var (
 		sers []map[string]string
-		err  error
+		lvc  = false // LabelValuesCall
```

Comment:
Nit: Would be cool to add same explanation above to `lnc` - looking first time at the code I needed a minute to connect the dots. Ideally even having full name `isLabelValueCall` instead of a comment would be nicer :nail_care:.

Reply:
Yea, I don't like those cryptic names

Reply:
Will fix